rest/middleware/cors.go

```diff
@@ -1,16 +1,97 @@
 package middleware
 
-import "net/http"
-
-// CORSMiddleware enables cross-origin requests. It implements the Middleware
-// interface.
-func CORSMiddleware(w http.ResponseWriter, r *http.Request) bool {
-	if origin := r.Header.Get("Origin"); origin != "" {
-		w.Header().Set("Access-Control-Allow-Origin", origin)
-		w.Header().Set("Access-Control-Allow-Methods", "POST, GET, OPTIONS, PUT, DELETE")
-		w.Header()["Access-Control-Allow-Headers"] = r.Header["Access-Control-Request-Headers"]
-		w.Header().Set("Access-Control-Allow-Credentials", "true")
+import (
+	"net/http"
+	"net/url"
+	"strings"
+
+	"github.com/Workiva/go-rest/rest"
+)
+
+type corsError struct {
+	msg  string
+	code int
+}
+
+func (c *corsError) Code() int {
+	return c.code
+}
+
+func (c *corsError) Response() []byte {
+	return []byte(c.msg)
+}
+
+func (c *corsError) Error() string {
+	return c.msg
+}
+
+// NewCORSMiddleware returns a Middleware which enables cross-origin requests.
+// Origin must match the supplied whitelist (which supports wildcards). Returns
+// a MiddlewareError if the request should be terminated.
+func NewCORSMiddleware(originWhitelist []string) rest.Middleware {
+	return func(w http.ResponseWriter, r *http.Request) rest.MiddlewareError {
+		originMatch := false
+		if origin := r.Header.Get("Origin"); checkOrigin(origin, originWhitelist) {
+			w.Header().Set("Access-Control-Allow-Origin", origin)
+			w.Header().Set("Access-Control-Allow-Methods", "POST, GET, OPTIONS, PUT, DELETE")
+			w.Header()["Access-Control-Allow-Headers"] = r.Header["Access-Control-Request-Headers"]
+			w.Header().Set("Access-Control-Allow-Credentials", "true")
+			originMatch = true
+		}
+
+		var err rest.MiddlewareError
+		if r.Method == "OPTIONS" {
+			err = &corsError{code: http.StatusOK}
+		}
+		if !originMatch {
```

Comment:
Do you care that your first error will be overwritten if both error conditions are true?